design/application-abi.md

```diff
@@ -0,0 +1,32 @@
+WASI Application ABI
+====================
+
+In addition to the APIs defined by the various WASI [modules](modules.md) there
+are also certain expectations that the WASI runtime places on an application
+that wishes to be portable across WASI implementations.
+
+This document describes how a conforming WASI application is expected to behave
+in terms of lifecycle (startup, shutdown, etc) and imports and exports the that
+it is expected to include.
+
+Lifecycle
+---------
+
+A WASI program may contain any number exports, but the embedded attributes
+specific meaning the following optional function exports:
+
+- `__wasi_init` - Called after WebAssemembly instantiation but before any other
+  functions.
+- `__wasi_main` - Runs the programs main entry points.  May be omitted, for
+  example, in the case of a library.
+- `__wasi_term` - The logical inverse of `__wasi_init`.  Optionally called
+  before module destruction.  No other functions within program will be called
+  after this one.
```

Comment:
I was also thinking along those lines.   I was basing this on the fact that we do currently have "post main" cleanup code in crt1.c:

```
    /* Call atexit functions, destructors, stdio cleanup, etc. */                
    __prepare_for_exit();    
```

However if we switch the Command model than such cleanup can stay where it as and run after main.  So for Commands yes I agree __wasi_term maybe is not needed.

For the case of Reactors would you not also want code like this (e.g. to flush stdout and run exit handler) in the case of a normal/ideal shutdown?

Reply:
I think this comes down to how Reactors will terminate. I imagine they'll terminate by calling `proc_exit` (or [whatever that ends up as](https://github.com/WebAssembly/WASI/issues/26)). In that case, they can do `__prepare_for_exit` or whatever other cleanup they need to do themselves at that time.

Reply:
You don't think there is any need for user/embedder of the Reactor to inform the Reactor when its don't using it?  Otherwise how does it know when it should prepare for its exit?

Another way of putting it:  `proc_exit` is something that a WASI application (the Reactor) would have when its exiting.     How do I as the user/embedder tell the reactor I'm done with it?

Reply:
I'm not sure about this, but one theory is that in a Reactor, there is no need for such a thing. Reactors shouldn't have data sitting around in a stdout buffer between events if they really care about it getting delivered, because in many contexts, there won't be a chance for a shutdown notification -- it'll be power-off-the-IoT-device, STONITH-the-server, WIFI-just-disappeared, abort-the-blockchain-transaction, or other things. I don't completely believe in this theory yet, but it would simplify some things.

I was going to add closer-the-browser-tab, but in that case, there is an `onbeforeunload` callback. If we want to expose the `onbeforeunload` callback to wasm via `__wasi_term`, we may want to adjust the wording to account for the fact that unloading can be cancelled and restarted, so perhaps `__wasi_term` may mean that no further functions will be called.

So, I'm ok adding `__wasi_term` here for now, and we an continue to think about this.